File: backend/pet-emarket-server/src/main/java/com/petemarket/server/product/Product.java
package com.petemarket.server.product;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;

@Entity
@Table(name = "product")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Long storeId = 1L;

    @Column(nullable = false, length = 128)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ProductType type = ProductType.GOODS;

    @Column(nullable = false, length = 64)
    private String category = "General";

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price = BigDecimal.ZERO;

    @Column(nullable = false)
    private Integer stock = 0;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ProductStatus status = ProductStatus.DRAFT;

    private String coverUrl;

    @Column(length = 2000)
    private String description;

    @Column(length = 64)
    private String petCode;

    @Column(length = 64)
    private String breed;

    @Column(length = 64)
    private String healthStatus;

    @Column(length = 128)
    private String vaccineCertNo;

    @Column(length = 128)
    private String quarantineCertNo;

    @Column(length = 255)
    private String traceSource;

    private Instant createdAt;
    private Instant updatedAt;

    @PrePersist
    void prePersist() {
        Instant now = Instant.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    void preUpdate() {
        updatedAt = Instant.now();
    }

    public Long getId() {
        return id;
    }

    public Long getStoreId() {
        return storeId;
    }

    public void setStoreId(Long storeId) {
        this.storeId = storeId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public ProductType getType() {
        return type;
    }

    public void setType(ProductType type) {
        this.type = type;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public Integer getStock() {
        return stock;
    }

    public void setStock(Integer stock) {
        this.stock = stock;
    }

    public ProductStatus getStatus() {
        return status;
    }

    public void setStatus(ProductStatus status) {
        this.status = status;
    }

    public String getCoverUrl() {
        return coverUrl;
    }

    public void setCoverUrl(String coverUrl) {
        this.coverUrl = coverUrl;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getPetCode() {
        return petCode;
    }

    public void setPetCode(String petCode) {
        this.petCode = petCode;
    }

    public String getBreed() {
        return breed;
    }

    public void setBreed(String breed) {
        this.breed = breed;
    }

    public String getHealthStatus() {
        return healthStatus;
    }

    public void setHealthStatus(String healthStatus) {
        this.healthStatus = healthStatus;
    }

    public String getVaccineCertNo() {
        return vaccineCertNo;
    }

    public void setVaccineCertNo(String vaccineCertNo) {
        this.vaccineCertNo = vaccineCertNo;
    }

    public String getQuarantineCertNo() {
        return quarantineCertNo;
    }

    public void setQuarantineCertNo(String quarantineCertNo) {
        this.quarantineCertNo = quarantineCertNo;
    }

    public String getTraceSource() {
        return traceSource;
    }

    public void setTraceSource(String traceSource) {
        this.traceSource = traceSource;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }
}

File: frontend/user-app/lib/models/product.dart
class Product {
  const Product({
    required this.id,
    required this.name,
    required this.type,
    required this.category,
    required this.price,
    required this.stock,
    required this.status,
    this.storeId = '',
    this.description = '',
    this.coverUrl = '',
    this.tags = const [],
    this.livePet,
  });

  final String id;
  final String name;
  final String type;
  final String category;
  final double price;
  final int stock;
  final String status;
  final String storeId;
  final String description;
  final String coverUrl;
  final List<String> tags;
  final Map<String, dynamic>? livePet;

  bool get isLivePet => type == 'PET_LIVE';

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      type: json['type']?.toString() ?? 'GOODS',
      category: json['category']?.toString() ?? 'General',
      price: NumberParser.toDouble(json['price']),
      stock: NumberParser.toInt(json['stock']),
      status: json['status']?.toString() ?? 'DRAFT',
      storeId: json['storeId']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      coverUrl: json['coverUrl']?.toString() ?? '',
      tags: (json['tags'] is List) ? (json['tags'] as List).map((item) => item.toString()).toList() : const [],
      livePet: _livePetFromJson(json),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type,
      'category': category,
      'price': price,
      'stock': stock,
      'status': status,
      'storeId': storeId,
      'description': description,
      'coverUrl': coverUrl,
      'tags': tags,
      'livePet': livePet,
    };
  }
}

Map<String, dynamic>? _livePetFromJson(Map<String, dynamic> json) {
  if (json['livePet'] is Map) {
    return Map<String, dynamic>.from(json['livePet'] as Map);
  }
  if (json['type']?.toString() != 'PET_LIVE') {
    return null;
  }
  return {
    'petCode': json['petCode']?.toString() ?? '',
    'breed': json['breed']?.toString() ?? '',
    'healthStatus': json['healthStatus']?.toString() ?? '',
    'vaccineCertNo': json['vaccineCertNo']?.toString() ?? '',
    'quarantineCertNo': json['quarantineCertNo']?.toString() ?? '',
    'traceSource': json['traceSource']?.toString() ?? '',
  };
}

class NumberParser {
  static double toDouble(Object? value) {
    if (value is num) return value.toDouble();
    return double.tryParse(value?.toString() ?? '') ?? 0;
  }

  static int toInt(Object? value) {
    if (value is num) return value.toInt();
    return int.tryParse(value?.toString() ?? '') ?? 0;
  }
}
